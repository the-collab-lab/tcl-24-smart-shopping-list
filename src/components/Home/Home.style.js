import { colorPalet } from '../Style/Color.palette.js';
import { Button } from '../Style/Button.Style.js';
import { Input } from '../Style/Input.Style.js';
import styled from 'styled-components';
import { device } from '../Style/Breakpoints.js';

const { secundary, bgApp, active, dark } = colorPalet;
const { sm } = device;

export const HomeContainer = styled.section`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  min-width: 375px;
  width: 100%;
  height: 90vh;
  background: ${bgApp};
  @media ${sm} {
    width: 420px;
    min-width: 420px;
    margin: auto;
    border-radius: 2rem;
    height: 80vh;
    box-shadow: 0 0px 6px rgba(0, 0, 0, 0.2);
  }
`;

export const TopSection = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-evenly;
  width: 100%;
  height: 55%;
  h1.TopSection-title {
    text-align: center;
    font-size: 3rem;
    text-transform: uppercase;
    width: 100%;
    height: 10%;
    color: ${dark};
    font-family: 'Josefin Sans', sans-serif;
  }
  img.TopSection-image {
    max-width: 100%;
    width: 20%;
    height: auto;
    margin-top: 2%;
  }
  p.TopSection-label {
    display: flex;
    align-items: center;
    height: 10%;
    font-size: 1.5rem;
    font-weight: bold;
    color: ${dark};
  }
`;

export const BottonSection = styled.div`
  width: 100%;
  height: 45%;
  background: ${dark};
  border-top-right-radius: 50px;
  border-top-left-radius: 50px;
  box-shadow: 0 0px 6px rgba(0, 0, 0, 0.2);
  @media ${sm} {
    border-bottom-left-radius: 2rem;
    border-bottom-right-radius: 2rem;
  }
`;

export const Form = styled.form`
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: center;
  width: 100%;
  height: 60%;
  margin: auto;
  border-top-left-radius: 40px;
  border-top-right-radius: 40px;
  label.BottonSection-Form-label {
    display: flex;
    flex-direction: column;
    justify-content: space-evenly;
    align-items: center;
    height: 50%;
    width: 100%;
  }
  p.BottonSection-Form-title {
    margin-bottom: 1%;
    font-size: 1.5rem;
    font-weight: bold;
    color: ${bgApp};
  }
`;

export const Notification = styled.div`
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 40%;
  margin: auto;
  color: ${bgApp};
  p {
    position: absolute;
    top: 0;
    font-size: 1.5rem;
    font-weight: bold;
  }

  .progress-3 {
    position: absolute;
    top: 0;
    width: 120px;
    height: 20px;
    border-radius: 20px;
    background: repeating-linear-gradient(
          135deg,
          #837a75 0 10px,
          #dafeb7 0 20px
        )
        left/0% 100% no-repeat,
      repeating-linear-gradient(135deg, #ddd 0 10px, #eee 0 20px) left/100% 100%;
    animation: p3 2s infinite;
  }
  @keyframes p3 {
    100% {
      background-size: 100% 100%;
    }
  }
`;

export const ButtonHome = styled(Button)`
  background: ${dark};
  font-size: 1rem;
  font-weight: bolder;
  text-transform: uppercase;
  color: ${bgApp};
  padding: 1.2rem 0.8rem;
  &:hover {
    background: ${secundary};
    color: ${active};
  }
`;

export const InputHome = styled(Input)`
  position: relative;
  width: 70%;
  height: 3.2rem;
  border-radius: 100px 100px 100px 100px;
  font-size: 1.2rem;
  background-color: #f8fdef;
  border: 0.1rem solid ${dark};
`;

export const ButtonHomeSearch = styled(Button)`
  position: absolute;
  top: 51%;
  left: 75%;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  background: none;
  box-shadow: none;
  i {
    font-size: 1.8rem;
    color: ${dark};
  }
  &:hover {
    background: none;
  }
`;
