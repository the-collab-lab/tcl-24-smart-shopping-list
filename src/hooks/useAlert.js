import { useEffect, useRef, useState } from 'react';

export const useAlert = (initialAlert = { type: '', message: '' }) => {
  const [alertMessage, setAlertMessage] = useState(initialAlert);
  const timeoutIdRef = useRef();

  const setTimeoutAlert = () => {
    timeoutIdRef.current = setTimeout(() => {
      setAlertMessage(initialAlert);
    }, 2000);
  };

  useEffect(() => {
    return () => {
      if (timeoutIdRef.current) clearTimeout(timeoutIdRef.current);
    };
  }, []);

  return [alertMessage, setTimeoutAlert, setAlertMessage];
};
