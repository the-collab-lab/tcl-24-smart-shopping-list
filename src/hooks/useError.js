import { useState } from 'react';

const useError = () => {
  const [load, setLoad] = useState('');
  const [error, setError] = useState('');
  const [success, setSuccess] = useState('');
  return {
    load,
    setLoad,
    error,
    setError,
    success,
    setSuccess,
  };
};

export default useError;
