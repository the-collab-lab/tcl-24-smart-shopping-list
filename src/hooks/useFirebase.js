import { fb } from './../lib/firebase';

export const useFirebase = () => {
  const db = fb.firestore().collection('lists');

  // const docRef = db.collection("lists").doc(
  //   'LW3QWM4IBucYC3tVDUiy'
  // );

  // console.log(docRef);

  const getAll = () => db;

  const getById = (id) => {
    return db.doc(id).get();
  };

  const create = (data) => {
    return db.doc(data.userToken).set(data);
  };

  const update = (id, value) => {
    return db.doc(id).update(value);
  };

  const remove = (id) => {
    return db.doc(id).delete();
  };

  return {
    getAll,
    getById,
    create,
    update,
    remove,
  };
};
