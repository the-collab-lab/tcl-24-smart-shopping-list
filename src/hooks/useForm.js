import { useState } from 'react';

export const useForm = (initialState = {}) => {
  const [values, setValues] = useState(initialState);

  const handleInputChange = ({ target }) => {
    setValues({ ...values, [target.name]: target.value });
  };

  const resetValues = (newFormState = initialState) => {
    setValues(newFormState);
  };

  return [values, handleInputChange, setValues, resetValues];
};
