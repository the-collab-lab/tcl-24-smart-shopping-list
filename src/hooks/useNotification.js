import { useState } from 'react';

const useNotification = () => {
  const [load, setLoad] = useState('');
  const [error, setError] = useState('');
  const [success, setSuccess] = useState('');
  return {
    load,
    setLoad,
    error,
    setError,
    success,
    setSuccess,
  };
};

export default useNotification;
