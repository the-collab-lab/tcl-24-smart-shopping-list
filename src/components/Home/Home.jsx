import { useHistory } from 'react-router-dom';
// import getToken from '../../lib/tokens';
import { useForm } from '../../hooks/useForm';
import { useFirebase } from '../../hooks/useFirebase.js';
import useNotification from '../../hooks/useNotification';
import { ArchivalNoticeModal } from '@the-collab-lab/shopping-list-utils';

import {
  HomeContainer,
  TopSection,
  BottonSection,
  ButtonHome,
  Form,
  InputHome,
  ButtonHomeSearch,
  Notification,
} from './Home.style.js';

import store from './Image/conifer-481.png';

const Home = () => {
  const [values, handleInputChange, setValues] = useForm({
    token: '',
  });

  const { load, setLoad, setError, error } = useNotification();

  const history = useHistory();
  const { getAll } = useFirebase();

  function searching() {
    setLoad('Searching...');

    const docRef = getAll().doc(values.token);

    docRef
      .get()
      .then((doc) => {
        if (doc.exists) {
          setLoad('');
          localStorage.setItem('token', values.token);
          history.push('/list');
        } else {
          setError('The list is not found');
          setLoad('');
          setTimeout(() => {
            setError('');
          }, 2000);
        }
      })
      .catch((error) => {
        setError(`Error getting document: ${error}`);
      });
  }

  const handleClick = () => {
    console.log('Creating new lists is disabled');
    // const token = getToken();
    // localStorage.setItem('token', token);
    // history.push('/list');
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    searching();
    setValues({
      token: '',
    });
  };

  return (
    <HomeContainer>
      <TopSection>
        <h1 className="TopSection-title">Smart Shopping App</h1>

        <img className="TopSection-image" src={store} alt="store" />

        <p className="TopSection-label">Get started by creating a new list.</p>

        <ButtonHome onClick={handleClick} data-testid="addList">
          New list
        </ButtonHome>
      </TopSection>

      <BottonSection>
        <Form onSubmit={handleSubmit}>
          <label className="BottonSection-Form-label">
            <p className="BottonSection-Form-title">
              Join an existing shopping list.
            </p>
            {
              <InputHome
                data-testid="input"
                type="text"
                name="token"
                onChange={handleInputChange}
                value={values.token}
                placeholder="Enter a token"
                required
              />
            }
          </label>

          <ButtonHomeSearch type="submit" aria-label="search list" primary>
            <i className="fas fa-search"></i>
          </ButtonHomeSearch>
        </Form>

        <Notification>
          {error && <p>{error}</p>}
          {load && <div className="progress-3"></div>}
        </Notification>
      </BottonSection>
      <ArchivalNoticeModal />
    </HomeContainer>
  );
};

export default Home;
