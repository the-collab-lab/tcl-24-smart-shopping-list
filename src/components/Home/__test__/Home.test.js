import { shallow } from 'enzyme';
import { useFirebase } from '../../../hooks/useFirebase';
import { useForm } from '../../../hooks/useForm';
import useNotification from '../../../hooks/useNotification';
import Home from '../Home';

jest.mock('../../../hooks/useForm');
jest.mock('../../../hooks/useNotification');
jest.mock('../../../hooks/useFirebase');

describe('Home', () => {
  it('deberia renderisar', () => {
    const values = { token: '' };
    const handleInputChange = jest.fn();
    const reset = jest.fn();

    useForm.mockReturnValue([values, handleInputChange, reset]);

    const load = '';
    const setLoad = jest.fn();
    const error = '';
    const setError = jest.fn();

    useNotification.mockReturnValue({ load, setLoad, error, setError });

    const getAll = jest.fn();

    useFirebase.mockReturnValue({ getAll });

    const wrapper = shallow(<Home />);

    expect(wrapper).toMatchSnapshot();
  });

  it('debe llamar el handle input change y limpiar la caja de texto', () => {
    const values = { token: '' };
    const handleInputChange = jest.fn();
    const reset = jest.fn();

    useForm.mockReturnValue([values, handleInputChange, reset]);

    const load = '';
    const setLoad = jest.fn();
    const error = '';
    const setError = jest.fn();

    useNotification.mockReturnValue({ load, setLoad, error, setError });

    const getAll = jest.fn(() => ({
      doc: jest.fn(() => ({
        get: jest.fn().mockImplementation(() => Promise.resolve(true)),
      })),
    }));

    useFirebase.mockReturnValue({ getAll });

    const wrapper = shallow(<Home />);
    const value = '123123';

    wrapper.find('input').simulate('change', { target: { value } });
    wrapper.find('form').simulate('submit', { preventDefault() {} });

    expect(handleInputChange).toHaveBeenCalled();
    expect(wrapper.find('input').prop('value')).toBe('');
  });
});
