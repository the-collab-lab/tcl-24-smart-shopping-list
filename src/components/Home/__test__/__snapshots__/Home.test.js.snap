// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home deberia renderisar 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <label>
      Type Your Token
      <input
        name="token"
        onChange={[MockFunction]}
        required={true}
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Search
      </button>
    </label>
  </form>
  <button
    onClick={[Function]}
  >
    New List
  </button>
</div>
`;
